halm4d/issue3/batchwriter/writer/batchwriter.go

```diff
@@ -0,0 +1,81 @@
+package writer
+
+import (
+	"log"
+	"os"
+)
+
+type BatchWriter struct {
+	buffer       []string
+	maxBatchSize int
+	outputFile   *os.File
+	writtenLines int
+}
+
+func removeAndCreateFile(outputFile string) (file *os.File, err error) {
+	if _, err := os.Stat(outputFile); err == nil {
+		err := os.Remove(outputFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+	file, err = os.Create(outputFile)
+	if err != nil {
+		return file, err
+	}
+	return file, err
+}
+
+func NewBatchWriter(maxBatchSize int, outputFilePath string, reCreateOutputFile bool) *BatchWriter {
+	var outputFile *os.File
+	if reCreateOutputFile {
+		var err error
+		outputFile, err = removeAndCreateFile(outputFilePath)
+		if err != nil {
+			log.Fatalln("Error recreating output file:", err)
+		}
+	}
+	return &BatchWriter{
+		buffer:       make([]string, 0),
+		maxBatchSize: maxBatchSize,
+		outputFile:   outputFile,
+	}
+}
+
+func (bw *BatchWriter) Write(s string) error {
+	bw.buffer = append(bw.buffer, s)
+	if len(bw.buffer) >= bw.maxBatchSize {
+		return bw.Flush()
+	}
+	return nil
+}
+
+func (bw *BatchWriter) Flush() (err error) {
+	file, err := os.OpenFile(bw.outputFile.Name(), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 06666)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = file.Close()
```

Comment:
actually, here, it would be even better:

```suggestion
		if err == nil {
		     err = file.Close()
		}
```

https://goplay.followthepattern.net/snippet/jJ3KQEmuFVG

Reply:
In this case the `file.Close()` won't be called when there is another error and the file stays open? What if I use `errors.Join()`?

Reply:
Removed this part of code, because of the refactor.